### ISSUE TITLE
Private sessions are not deleted when settings.closePrivateTabs is set

### ISSUE THREAD

**Issue by michaelsproul** [mozilla-mobile/firefox-ios#5634]
When leaving a private browsing session with the "Close Private Tabs" setting enabled, the session _is not_ deleted as you would expect, and will be present the next time you enter private browsing.

## Steps to reproduce

1. Enable "Close Private Tabs" (i.e. `settings.closePrivateTabs`)
2. Open a private browsing session, and search for something in the URL/search bar
3. Leave the private browsing session, either by closing the tab or tapping the purple mask icon
4. Now that you're back in the normal session again, open a _new_ private browsing session. The tab from before will be gone. Search for a new term in the URL/search bar
5. Once the results page loads, delete the query from Google's search bar and notice that the query from step (2) is listed in the search history! 😱

## Screenshot

<img src="https://i.imgur.com/VdVB7CX.png" width="400px"/>

## Build info

Firefox for iOS, Version 19.1 (16203)

**Comment by michaelsproul** [comment]
Still present in Version 20.0 (16437)

**Comment by garvankeeley** [comment]
I can't repro this so far. What OS version are you on?

**Comment by michaelsproul** [comment]
iOS 12.4.2 on an iPhone 6

**Comment by SimonBasca** [comment]
I was able to reproduce this issue on 20.0 (16473) iPhone 7 (12.31) following the STR.

1. Go to Private browsing 
2. Tap on the URL bar and search for "query 1"
3. Long press on the  Tab Counter and tap on "Close Tab"
4. Long press on the  Tab Counter and tap on "Private Browsing"
5 Tap on the URL bar and search for "query 2"
6. Tap on the Google search bar and clear the text.

Note:
- This is not reproducible on iOS 13.
- This is also reproducible without enabling "Close Private Tabs".
- Restarting the browser will delete the session correctly and the previous searches will be cleared.
 
Video:
- [Issue.zip](https://github.com/mozilla-mobile/firefox-ios/files/3797456/Issue.zip)

**Comment by garvankeeley** [comment]
Confirmed it seems to be iOS 12.x only. Didn't happen on 11.x or 13.x.

**Comment by nbhasin2** [comment]
This seems to be amalgam of bug b/w iOS 12.x side and the app. When you have multiple tabs open and if you have an instance of even a single webview available in the background this bug is caused.
The only way to avoid this bug is to only have a single instance of a webview (at-least that's the case during my investigation). Here are few things I have tried:

a) Clear webview data in private mode 
--- We already do this but I tried clearing just the google data as well and still didn't work
b) Clear all the data in the WKWebsiteDataStore 
--- This works but only when the app is not in private mode. As soon as we try to clear it while in private mode the bug still exist. 
c) Make a dummy app and see if clearing the data on the dummy app works on iOS 12.x
--- It does clear the webview data (for single WKWebView) while using the following methods 

```
let types = WKWebsiteDataStore.allWebsiteDataTypes()
WKWebsiteDataStore.default().removeData(ofTypes: types, modifiedSince: .distantPast) {}
---
let dataTypes = WKWebsiteDataStore.allWebsiteDataTypes()

webView?.configuration.websiteDataStore.removeData(ofTypes: dataTypes,
                                                     modifiedSince: Date.distantPast,
                                                 completionHandler: completionHandler)
```

What else can be done:
We can try to dig in how the history is being stored on the disk or memory in the app and see if that's causing it?

Keep in mind that this bug is not present in iOS 13 with no change to the app.
